e-highways connection problems in full workflow

### What happened?

Running the full workflow, the ehighways resolution failed in rule `link_locations_with_transmission_techs_template`. 
See log below.

### Version

1.2.0.dev

### Relevant log output

```shell
[Tue Jun 11 12:44:46 2024]
Job 406: Link ehighways direct neighbours and neighbours with sea connections with transmission techs from template.
Reason: Missing output files: build/models/ehighways/techs/transmission/electricity-linked-neighbours.yaml; Updated input files: build/data/ehighways/units.geojson

Activating conda environment: ../envs/snakemake/euro-calliope/afba709282c28f1f96f3b898cb927b2e_
Activating conda environment: ../envs/snakemake/euro-calliope/afba709282c28f1f96f3b898cb927b2e_
Traceback (most recent call last):
  File "/Users/adrienmellot/Documents/GitHub/euro-calliope/.snakemake/scripts/tmp0axpx5qw.template_transmission.py", line 70, in <module>
    construct_techs_and_links(
  File "/Users/adrienmellot/Documents/GitHub/euro-calliope/.snakemake/scripts/tmp0axpx5qw.template_transmission.py", line 27, in construct_techs_and_links
    assert nx.is_connected(graph), "There are electrical islands in the network graph."
AssertionError: There are electrical islands in the network graph.
RuleException:
CalledProcessError in file /Users/adrienmellot/Documents/GitHub/euro-calliope/./rules/transmission.smk, line 52:
Command 'source /Users/adrienmellot/miniconda3/bin/activate '/Users/adrienmellot/Documents/GitHub/envs/snakemake/euro-calliope/afba709282c28f1f96f3b898cb927b2e_'; set -euo pipefail;  python /Users/adrienmellot/Documents/GitHub/euro-calliope/.snakemake/scripts/tmp0axpx5qw.template_transmission.py' returned non-zero exit status 1.
[Tue Jun 11 12:44:48 2024]
Error in rule link_locations_with_transmission_techs_template:
    jobid: 406
    input: templates/models/techs/transmission/electricity-transmission.yaml, build/data/ehighways/units.geojson
    output: build/models/ehighways/techs/transmission/electricity-linked-neighbours.yaml
    conda-env: /Users/adrienmellot/Documents/GitHub/envs/snakemake/euro-calliope/afba709282c28f1f96f3b898cb927b2e_

Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: .snakemake/log/2024-06-11T124347.887384.snakemake.log
WorkflowError:
At least one job did not complete successfully.
```

--- a/scripts/shapes/units.py
+++ b/scripts/shapes/units.py
@@ -52,7 +52,7 @@ def _build_layer(
         for country, source_layer in resolution_config.items()
     ])
     assert isinstance(layer, pd.DataFrame)
-    return gpd.GeoDataFrame(layer, crs=crs)
+    return gpd.GeoDataFrame(layer, crs=crs).reset_index(drop=True)
 
 
 def _read_source_layers(
@@ -121,7 +121,15 @@ def _rename_ehighways_countries(units: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
     Method is future-proofed to handle higher resolution than country code (e.g. using GADM3 units in one country).
     It does this by ordering all non-ehighways IDs alphabetically and then turning them into "XXX_#" style codes.
     """
-    sorted_units = units.sort_values(by="id")
+    order = (
+        units.id.str.extractall(r"(\d+)")
+        .groupby(level=0)
+        .sum()
+        .reindex(units.index)
+        .fillna(1)
+        .astype(int)
+    )
+    sorted_units = units.assign(order=order).sort_values(by=["country_code", "order"])
 
     new_names = (
         sorted_units.country_code
@@ -132,7 +140,8 @@ def _rename_ehighways_countries(units: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
     sorted_units["id"] = sorted_units.id.where(
         sorted_units.type == "ehighways", new_names
     )
-    return sorted_units
+
+    return sorted_units.drop("order", axis=1)
 
 
 if __name__ == "__main__":

--- a/scripts/shapes/units.py
+++ b/scripts/shapes/units.py
@@ -45,14 +45,12 @@ def _build_layer(
         resolution_config
     """
     crs = [layer.crs for layer in source_layers.values()][0]
-    layer = pd.concat(
-        [
-            source_layers[source_layer][
-                source_layers[source_layer].country_code == _iso3(country)
-            ]
-            for country, source_layer in resolution_config.items()
+    layer = pd.concat([
+        source_layers[source_layer][
+            source_layers[source_layer].country_code == _iso3(country)
         ]
-    )
+        for country, source_layer in resolution_config.items()
+    ])
     assert isinstance(layer, pd.DataFrame)
     return gpd.GeoDataFrame(layer, crs=crs).reset_index(drop=True)
 
@@ -85,17 +83,15 @@ def _validate_resolution_config(
 ):
     missing_countries = set(all_countries).difference(resolution_config)
 
-    assert (
-        not missing_countries
-    ), f"Missing countries in {resolution} resolution configuration: {missing_countries}"
+    assert not missing_countries, f"Missing countries in {resolution} resolution configuration: {missing_countries}"
 
 
 def _verify_all_countries_captured(
     countries: list[str], units: gpd.GeoDataFrame, resolution: str
 ):
-    missing_countries = set(units.country_code).difference(
-        [_iso3(country) for country in countries]
-    )
+    missing_countries = set(units.country_code).difference([
+        _iso3(country) for country in countries
+    ])
     assert (
         not missing_countries
     ), f"Countries are missing in {resolution} resolution shapes: {missing_countries}"

--- a/scripts/shapes/units.py
+++ b/scripts/shapes/units.py
@@ -45,12 +45,14 @@ def _build_layer(
         resolution_config
     """
     crs = [layer.crs for layer in source_layers.values()][0]
-    layer = pd.concat([
-        source_layers[source_layer][
-            source_layers[source_layer].country_code == _iso3(country)
+    layer = pd.concat(
+        [
+            source_layers[source_layer][
+                source_layers[source_layer].country_code == _iso3(country)
+            ]
+            for country, source_layer in resolution_config.items()
         ]
-        for country, source_layer in resolution_config.items()
-    ])
+    )
     assert isinstance(layer, pd.DataFrame)
     return gpd.GeoDataFrame(layer, crs=crs).reset_index(drop=True)
 
@@ -83,15 +85,17 @@ def _validate_resolution_config(
 ):
     missing_countries = set(all_countries).difference(resolution_config)
 
-    assert not missing_countries, f"Missing countries in {resolution} resolution configuration: {missing_countries}"
+    assert (
+        not missing_countries
+    ), f"Missing countries in {resolution} resolution configuration: {missing_countries}"
 
 
 def _verify_all_countries_captured(
     countries: list[str], units: gpd.GeoDataFrame, resolution: str
 ):
-    missing_countries = set(units.country_code).difference([
-        _iso3(country) for country in countries
-    ])
+    missing_countries = set(units.country_code).difference(
+        [_iso3(country) for country in countries]
+    )
     assert (
         not missing_countries
     ), f"Countries are missing in {resolution} resolution shapes: {missing_countries}"
@@ -121,6 +125,8 @@ def _rename_ehighways_countries(units: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
     Method is future-proofed to handle higher resolution than country code (e.g. using GADM3 units in one country).
     It does this by ordering all non-ehighways IDs alphabetically and then turning them into "XXX_#" style codes.
     """
+    # Extract all digits from a string, merge them in order via string summation, then convert to an integer
+    # E.g., "BE.1.2-3" -> 123
     order = (
         units.id.str.extractall(r"(\d+)")
         .groupby(level=0)

